src/mds/MDLog.cc

```diff
@@ -291,30 +318,29 @@ void MDLog::_submit_entry(LogEvent *le, MDSLogContextBase* c)
   unflushed++;
 }
 
-void MDLog::_segment_upkeep(LogEvent* le)
+void MDLog::_segment_upkeep()
 {
   ceph_assert(ceph_mutex_is_locked_by_me(mds->mds_lock));
   ceph_assert(ceph_mutex_is_locked_by_me(submit_mutex));
   uint64_t period = journaler->get_layout_period();
   auto ls = get_current_segment();
   auto log_events_per_segment = g_conf()->mds_log_events_per_segment;
+  auto major_segment_event_ratio = g_conf()->mds_major_segment_event_ratio;
   // start a new segment?
-  if (le->get_type() == EVENT_SUBTREEMAP ||
-      (le->get_type() == EVENT_IMPORTFINISH && mds->is_resolve())) {
-    // avoid infinite loop when ESubtreeMap is very large.
-    // do not insert ESubtreeMap among EImportFinish events that finish
-    // disambiguate imports. Because the ESubtreeMap reflects the subtree
-    // state when all EImportFinish events are replayed.
+  if (events_since_last_major_segment > log_events_per_segment*major_segment_event_ratio) {
+    dout(10) << __func__ << ": starting new major segment, current " << *ls << dendl;
+    auto sle = mds->mdcache->create_subtree_map();
+    _submit_entry(sle, NULL);
   } else if (ls->end/period != ls->offset/period || ls->num_events >= log_events_per_segment) {
-    dout(10) << "submit_entry also starting new segment: last = "
-            << ls->seq  << "/" << ls->offset << ", event seq = " << event_seq << dendl;
-    _start_new_segment();
-  } else if (g_conf()->mds_debug_subtrees && le->get_type() != EVENT_SUBTREEMAP_TEST) {
+    dout(10) << __func__ << ": starting new segment, current " << *ls << dendl;
+    auto sb = new ESegment();
```

Comment:
It's not 100% clear to me about the usage/introduction of ESegment event. I understand that its a minor segment (boundary) and starts a new log segment replacing ESubtreeMap. So, it seems that this event type is mostly to satisfy the presence (beginning) of a log segment? The deferment of ESubtreeMap journal event is what's going to provide the real benefits?

Reply:
> So, it seems that this event type is mostly to satisfy the presence (beginning) of a log segment?

Yes.

> The deferment of ESubtreeMap journal event is what's going to provide the real benefits?

Yes.